File: vane-core/src/main/java/org/oddlama/vane/core/command/params/AnyParam.java
package org.oddlama.vane.core.command.params;

import java.util.Collections;
import java.util.List;
import org.bukkit.command.CommandSender;
import org.oddlama.vane.core.command.Command;
import org.oddlama.vane.core.command.check.CheckResult;
import org.oddlama.vane.core.command.check.ErrorCheckResult;
import org.oddlama.vane.core.command.check.ParseCheckResult;
import org.oddlama.vane.core.functional.Function1;

public class AnyParam<T> extends BaseParam {

    private String argumentType;
    private Function1<String, ? extends T> fromString;

    public AnyParam(Command<?> command, String argumentType, Function1<String, ? extends T> fromString) {
        super(command);
        this.argumentType = argumentType;
        this.fromString = fromString;
    }

    @Override
    public CheckResult checkParse(CommandSender sender, String[] args, int offset) {
        if (args.length <= offset) {
            return new ErrorCheckResult(offset, "§6missing argument: §3" + argumentType + "§r");
        }
        var parsed = parse(args[offset]);
        if (parsed == null) {
            return new ErrorCheckResult(offset, "§6invalid §3" + argumentType + "§6: §b" + args[offset] + "§r");
        }
        return new ParseCheckResult(offset, argumentType, parsed, true);
    }

    @Override
    public List<String> completionsFor(CommandSender sender, String[] args, int offset) {
        return Collections.emptyList();
    }

    private T parse(String arg) {
        return fromString.apply(arg);
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/command/params/AnyParam.kt
package org.oddlama.vane.core.command.params

import org.bukkit.command.CommandSender
import org.oddlama.vane.core.command.Command
import org.oddlama.vane.core.command.check.CheckResult
import org.oddlama.vane.core.command.check.ErrorCheckResult
import org.oddlama.vane.core.command.check.ParseCheckResult
import org.oddlama.vane.core.functional.Function1

/**
 * Parameter that accepts any argument parsed by a converter function.
 *
 * @param T the parsed argument type.
 * @param command the owning command.
 * @param argumentType display name used in usage and error messages.
 * @param fromString parser that converts raw argument strings.
 */
class AnyParam<T>(
    command: Command<*>?,
    /** Argument type name used in errors and usage. */
    private val argumentType: String,
    /** Parser converting raw strings into typed values. */
    private val fromString: Function1<String?, out T?>
) : BaseParam(command) {

    /**
     * Parses this argument and returns either a parse result or an error.
     */
    override fun checkParse(sender: CommandSender?, args: Array<String?>?, offset: Int): CheckResult {
        if (args == null || args.size <= offset)
            return ErrorCheckResult(offset, "§6missing argument: §3$argumentType§r")
        val parsed = fromString.apply(args[offset])
            ?: return ErrorCheckResult(offset, "§6invalid §3$argumentType§6: §b${args[offset]}§r")
        return ParseCheckResult(offset, argumentType, parsed, true)
    }

    /**
     * Returns completions for this parameter.
     */
    override fun completionsFor(sender: CommandSender?, args: Array<String?>?, offset: Int): MutableList<String?> =
        mutableListOf()
}
